File: src/main/java/uk/me/mantas/eternity/handlers/ImportCharacter.java
/**
 *  Eternity Keeper, a Pillars of Eternity save game editor.
 *  Copyright (C) 2015 the authors.
 *
 *  Eternity Keeper is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  Eternity Keeper is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


package uk.me.mantas.eternity.handlers;

import org.cef.browser.CefBrowser;
import org.cef.callback.CefQueryCallback;
import org.cef.callback.CefRunFileDialogCallback;
import org.cef.handler.CefDialogHandler.FileDialogMode;
import org.cef.handler.CefMessageRouterHandlerAdapter;
import org.json.JSONException;
import org.json.JSONObject;
import uk.me.mantas.eternity.Logger;
import uk.me.mantas.eternity.environment.Environment;
import uk.me.mantas.eternity.save.CharacterImporter;
import uk.me.mantas.eternity.save.CharacterImporter.ImportConflict;
import uk.me.mantas.eternity.save.SavedGameOpener;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Optional;
import java.util.Vector;

public class ImportCharacter extends CefMessageRouterHandlerAdapter {
	private static final Logger logger = Logger.getLogger(ImportCharacter.class);

	@Override
	public boolean onQuery (
		CefBrowser browser
		, long id
		, String request
		, boolean persistent
		, CefQueryCallback callback) {

		// Second phase of an import: the client already knows which CHR file
		// to use (typically after the user confirmed an overwrite), so no
		// file dialog is needed.
		String confirmedChrPath = null;
		boolean overwrite = false;
		try {
			final JSONObject json = new JSONObject(request);
			if (json.has("chrPath")) {
				confirmedChrPath = json.getString("chrPath");
				overwrite = json.optBoolean("overwrite", false);
			}
		} catch (final JSONException e) {
			logger.error("Error parsing JSON request: %s%n", request);
			callback.failure(-1, "Error parsing JSON request.");
			return true;
		}

		if (confirmedChrPath != null) {
			final String chrPath = confirmedChrPath;
			final boolean overwriteExisting = overwrite;
			Environment.getInstance().mutationWorker().execute(
				() -> doImport(request, callback, chrPath, overwriteExisting));
		} else {
			// The file dialog itself must not block the mutation queue; only
			// the import work moves there once a file has been chosen.
			Environment.getInstance().workers().execute(
				new SelectChrFile(browser, request, callback));
		}

		return true;
	}

	@Override
	public void onQueryCanceled (CefBrowser browser, long id) {
		logger.error("Query #%d cancelled.%n", id);
	}

	private class SelectChrFile implements Runnable {
		private final CefBrowser browser;
		private final String request;
		private final CefQueryCallback callback;

		public SelectChrFile (
			CefBrowser browser
			, String request
			, CefQueryCallback callback) {

			this.browser = browser;
			this.request = request;
			this.callback = callback;
		}

		@Override
		public void run () {
			browser.runFileDialog(
				FileDialogMode.FILE_DIALOG_OPEN
				, "Choose a character"
				, ""
				, new Vector<String>(){{add(".chr");}}
				, 0
				, new FileCallback(request, callback));
		}
	}

	private class FileCallback implements CefRunFileDialogCallback {
		private final String request;
		private final CefQueryCallback callback;

		public FileCallback (
			String request
			, CefQueryCallback callback) {

			this.request = request;
			this.callback = callback;
		}

		@Override
		public void onFileDialogDismissed (
			final int selectedAcceptFilter
			, final Vector<String> filenames) {

			if (filenames.size() < 1 || filenames.get(0).length() < 1) {
				callback.failure(-1, "NO_SAVE");
				return;
			}

			final String chrFile = filenames.get(0);
			Environment.getInstance().mutationWorker().execute(
				() -> analyseThenImport(request, callback, chrFile));
		}
	}

	// First phase: if the character already exists in the target save, ask
	// the client to confirm the overwrite instead of importing right away.
	private void analyseThenImport (
		final String request
		, final CefQueryCallback callback
		, final String chrFile) {

		try {
			final CharacterImporter importer = new CharacterImporter(request, chrFile);
			final Optional<ImportConflict> conflict = importer.detectConflict();

			if (conflict.isPresent()) {
				final JSONObject confirm = new JSONObject();
				confirm.put("characterName", conflict.get().characterName);
				confirm.put("isMainCharacter", conflict.get().mainCharacter);
				confirm.put("chrPath", chrFile);

				final JSONObject response = new JSONObject();
				response.put("confirmOverwrite", confirm);
				callback.success(response.toString());
				return;
			}
		} catch (final JSONException e) {
			logger.error("Error parsing JSON request: %s%n", request);
			callback.failure(-1, "Error parsing JSON request.");
			return;
		} catch (final FileNotFoundException e) {
			logger.error("File not found: %s%n", e.getMessage());
			callback.failure(-1, "Unable to find your save or CHR file.");
			return;
		} catch (final IOException e) {
			logger.error("%s%n", e.getMessage());
			callback.failure(-1, "Error reading save or CHR file.");
			return;
		}

		doImport(request, callback, chrFile, false);
	}

	private void doImport (
		final String request
		, final CefQueryCallback callback
		, final String chrFile
		, final boolean overwrite) {

		try {
			final CharacterImporter importer = new CharacterImporter(request, chrFile);
			final boolean success = overwrite
				? importer.overwriteCharacter()
				: importer.importCharacter();

			if (success) {
				final SavedGameOpener opener = new SavedGameOpener(
					importer.saveFile.getAbsolutePath()
					, callback);

				opener.run();
			} else {
				callback.failure(-1, "Character import failed.");
			}
		} catch (JSONException e) {
			logger.error("Error parsing JSON request: %s%n", request);
			callback.failure(-1, "Error parsing JSON request.");
		} catch (FileNotFoundException e) {
			logger.error("File not found: %s%n", e.getMessage());
			callback.failure(-1, "Unable to find your save or CHR file.");
		} catch (IOException e) {
			logger.error("%s%n", e.getMessage());
			callback.failure(-1, "Error modifying temporary MobileObjects.save");
		}
	}
}

File: src/ui/js/ImportCharacter.js
/**
 *  Eternity Keeper, a Pillars of Eternity save game editor.
 *  Copyright (C) 2015 the authors.
 *
 *  Eternity Keeper is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  Eternity Keeper is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

var ImportCharacter = function () {
	var self = this;

	var defaultState = {
		enabled: false
		, importing: false
	};

	self.state = $.extend({}, defaultState);
	self.html = {};

	// The path of the CHR file awaiting overwrite confirmation.
	var pendingChrPath = null;

	var setBusy = busy => {
		self.state.importing = busy;
		// Block the save button (and anything else keyed off `saving`) while
		// the import is writing the extracted save.
		Eternity.Modifications.transition({saving: busy});
	};

	var success = response => {
		setBusy(false);
		response = JSON.parse(response);

		if (response.error) {
			var msg;
			if (response.error === 'DESERIALIZATION_ERR') {
				msg = 'Error deserializing save file, please report your eternity.log file.';
			} else {
				msg = response.error;
			}

			Eternity.GenericError.render({msg: msg});
			return;
		}

		// The character already exists in this save; ask the user whether to
		// overwrite it before doing anything.
		if (response.confirmOverwrite) {
			pendingChrPath = response.confirmOverwrite.chrPath;

			var name = response.confirmOverwrite.characterName;
			var message = response.confirmOverwrite.isMainCharacter
				? 'Importing "' + name + '" will overwrite the existing main character in this'
					+ ' save file, including their stats and equipment.'
				: '"' + name + '" is already in this party. Importing them will overwrite the'
					+ ' existing character, including their stats and equipment.';

			self.html.importOverwriteDialog.find('.overwrite-message').text(message);
			self.html.importOverwriteDialog.modal('show');
			return;
		}

		// The importer modified the extracted save so we re-render the whole
		// save view with the fresh data and flag it as having modifications.
		Eternity.SavedGame.render({
			saveData: response
			, info: Eternity.SavedGame.state.info
		});

		Eternity.Modifications.transition({modifications: true});
	};

	var failure = (errno, response) => {
		setBusy(false);

		// NO_SAVE means the user cancelled the file dialog.
		if (response !== 'NO_SAVE') {
			Eternity.GenericError.render({msg: response});
		}
	};

	var baseRequest = () => {
		return {
			oldSave: Eternity.SavedGame.state.info.absolutePath
			, savedYet: Eternity.Modifications.state.savedYet
		};
	};

	self.init = () => {
		self.html.importOverwriteAgree.click(self.confirmOverwrite.bind(self));
		self.html.importOverwriteDialog.on('hidden.bs.modal', () => {
			pendingChrPath = null;
		});
	};

	self.importCharacter = () => {
		if (self.state.importing) {
			return;
		}

		setBusy(true);
		Eternity.GenericError.render({});

		window.importCharacter({
			request: JSON.stringify(baseRequest())
			, onSuccess: success
			, onFailure: failure
		});
	};

	self.confirmOverwrite = () => {
		var chrPath = pendingChrPath;
		self.html.importOverwriteDialog.modal('hide');

		if (chrPath === null || self.state.importing) {
			return;
		}

		var request = baseRequest();
		request.chrPath = chrPath;
		request.overwrite = true;

		setBusy(true);
		window.importCharacter({
			request: JSON.stringify(request)
			, onSuccess: success
			, onFailure: failure
		});
	};

	self.render = newState => {
		self.state = $.extend({}, defaultState, newState);
		var menuItem = self.html['menu-import-character'];
		menuItem.off();

		if (self.state.enabled) {
			menuItem.parent().removeClass('disabled');
			menuItem.click(self.importCharacter);
		} else {
			menuItem.parent().addClass('disabled');
		}
	};
};

$.extend(ImportCharacter.prototype, Renderer.prototype);
